game/dino_jump_main.py

```diff
@@ -0,0 +1,394 @@
+        self.WIDTH, self.HEIGHT = (
+            150,
+            48,
+        )  # define pixel height and width for actual board
+        self.GROUND_Y, self.LOW_BIRD_Y, self.HIGH_BIRD_Y = (
+            2,
+            18,
+            28,
+        )  # proportions for the three axes of obstacles
+
+        # Non-Editable Constants
+        self.BG_COLOR = (255, 255, 255)  # white colour (background)
+        self.BLOCK_COLOR = (0, 0, 0)  # black colour (sprites)
+        self.FPS = 60  # frames per second frames update
+        self.GROUND_AXIS = (
+            self.HEIGHT - self.GROUND_Y
+        )  # defines axis height with (0,0) lower left corner
+        self.LOW_BIRD_AXIS = (
+            self.HEIGHT - self.LOW_BIRD_Y
+        )  # defines axis height with (0,0) lower left corner
+        self.HIGH_BIRD_AXIS = (
+            self.HEIGHT - self.HIGH_BIRD_Y
+        )  # defines axis height with (0,0) lower left corner
+        self.start_game = False
+        self.game_over = False
+        self.close_all = False
+
+        # Setup Display
+        self.game_state_array = np.zeros((self.HEIGHT, self.WIDTH))
+        self.screen = pygame.display.set_mode(
+            (self.WIDTH * self.GRID_SIZE, self.HEIGHT * self.GRID_SIZE)
+        )
+        self.clk = pygame.time.Clock()
+        pygame.display.set_caption("DinoJump!")
+
+        # Initiate the Dino and ObstacleManager Classes
+        self.dino = Dino(self.GROUND_AXIS)
+        self.all_obstacles = ObstacleManager(
+            self.WIDTH, (self.GROUND_AXIS, self.LOW_BIRD_AXIS, self.HIGH_BIRD_AXIS)
+        )
+
+    def update_board_pixels(self, curr_object):
+
+        # Get all the values from the curr_object and the board
+        x, y = int(curr_object.x), int(curr_object.y)
+        height, width = curr_object.HEIGHT, curr_object.WIDTH
+        board_width = self.game_state_array.shape[1]
+        min_width, max_width = min(max(x, 0), board_width), max(
+            0, min(x + width, board_width)
+        )
+
+        # Segment the display section to draw to (especially if cropped)
+        left_edge = curr_object.ARRAY[:, -(max_width - min_width) :].astype(int)
+        right_edge = curr_object.ARRAY[:, : (max_width - min_width)].astype(int)
+        set_location = self.game_state_array[
+            y : y + height, min_width:max_width
+        ].astype(int)
+
+        # Extract bitwise AND and OR of incorporating the new drawing
+        if min_width == 0 and max_width != 0:
+            filled_location = np.bitwise_or(left_edge, set_location)
+            is_hit = np.sum(np.bitwise_and(left_edge, set_location))
+        else:
+            filled_location = np.bitwise_or(right_edge, set_location)
+            is_hit = np.sum(np.bitwise_and(right_edge, set_location))
+        self.game_state_array[y : y + height, min_width:max_width] = filled_location
+
+        # If a hit is detected, we end the game
+        if is_hit:
+            self.game_over = True

```

Comment:
This logic can probably be offloaded to a separate function purely for obstacle collision, and we can make the `update_board_pixels` function purely for updating the pixel locations

Reply:
Yes good point, they are seperated now, I added a background and full_game_state as separate frames for the collision comparison